contrib/babelfishpg_tsql/src/linked_servers.c

```diff
@@ -0,0 +1,1024 @@
+				
+				memcpy(&pbuf->data[1], numeric_bytes, 16);
+
+				pbuf->maxlen = 17;		/* sign byte + numeric bytes (1 + 16) */
+				pbuf->len = 17;
+				pbuf->cursor = 0;
+
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr(pbuf, TSQL_NUMERIC, numeric->scale);
+			}
+			break;
+		case TSQL_FLOATN:
+		case TSQL_FLOAT:
+			return Float8GetDatum(*(float8 *)val);
+		case TSQL_REAL:
+			return Float4GetDatum(*(float4 *)val);
+		case TSQL_TINYINT:
+			return UInt8GetDatum(*(int16_t *)val);
+		case TSQL_SMALLINT:
+			return Int16GetDatum(*(int16_t *)val);
+		case TSQL_INT:
+		case TSQL_INTN:
+			return Int32GetDatum(*(int32_t *)val);
+		case TSQL_BIGINT:
+			return Int64GetDatum(*(int64_t *)val);
+		case TSQL_MONEY:
+		case TSQL_MONEYN:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr)
+			{
+				StringInfo pbuf;
+
+				pbuf = palloc(sizeof(StringInfoData));
+				/*
+				 * Rather than copying data around, we just set up a phony
+				 * StringInfo pointing to the correct portion of the TDS message
+				 * buffer. 
+				 */
+				pbuf->data = (char *)val;
+				pbuf->maxlen = 8;
+				pbuf->len = 8;
+				pbuf->cursor = 0;
+
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr(pbuf, TSQL_MONEYN, 0);
+			}
+			break;
+		case TSQL_SMALLMONEY:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr)
+			{
+				StringInfo pbuf;
+
+				pbuf = palloc(sizeof(StringInfoData));
+				/*
+				 * Rather than copying data around, we just set up a phony
+				 * StringInfo pointing to the correct portion of the TDS message
+				 * buffer. 
+				 */
+				pbuf->data = (char *)val;
+				pbuf->maxlen = 4;
+				pbuf->len = 4;
+				pbuf->cursor = 0;
+
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_byte_ptr(pbuf, TSQL_SMALLMONEY, 0);
+			}
+			break;
+		case TSQL_DATETIME2:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_date_time_struct)
+			{
+				LS_TDS_DATETIMEALL *datetime2 = (LS_TDS_DATETIMEALL *) val;
+
+				/* optional attribute here is scale */
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_date_time_struct(datetime2->time, datetime2->date, TSQL_DATETIME2, 7);
+			}
+			break;
+		case TSQL_DATETIMEOFFSET:
+			if (*pltsql_protocol_plugin_ptr && (*pltsql_protocol_plugin_ptr)->get_datum_from_date_time_struct)
+			{
+				LS_TDS_DATETIMEALL *datetimeoffset = (LS_TDS_DATETIMEALL *) val;
+				
+				/* optional attribute here is time offset */
+				return (*pltsql_protocol_plugin_ptr)->get_datum_from_date_time_struct(datetimeoffset->time, datetimeoffset->date, TSQL_DATETIMEOFFSET, datetimeoffset->offset);
+			}
+			break;
+		default:
+			return (Datum) 0;
+	}
+
+	return (Datum) 0;
+}
+
+/*
+ * Given T-SQL data type in string, return equivalent client library TDS
+ * type. Used when preparing tuple descriptor for T-SQL OPENQUERY.
+ */
+int
+tdsTypeStrToTypeId(char* datatype)
+{
+	if (strcmp(datatype, "image") == 0)
```

Comment:
is it always lower case?

Reply:
Yes. I'm converting datatype to lowercase explicitly now just to be safe